cmd/cigarToBed/cigarToBed.go

```diff
@@ -0,0 +1,198 @@
+	}
+	numSeqOne := CountSeqIdx(inputFileOne)
+	numSeqTwo := CountSeqIdx(inputFileTwo)
+	if numSeqOne > 1 || numSeqTwo > 1 {
+		log.Fatalf("multiple sequnces detected in .fa files: %v sequences in the first .fa file and %v sequences in the second .fa file. This program is designed for .fa files with only 1 sequence in them\n", numSeqOne, numSeqTwo)
+	}
+	//attempt at Dan's len(faOne) suggestion
+	//if len(string(faOne)) != 1 || len(string(faTwo)) != 1 {
+	//log.Fatalf("multiple sequnces detected in .fa files. This program is designed for .fa files with only 1 sequence in them.\n")
+	//}
+
+	//needleman wunsch (global alignment)
+	//raven's note: scoring matrices are in align/align.go, e.g. HumanChimp matrix, same as what UCSC Genome Browser uses
+	//raven's note: ConstGap is a function in align/linearGap.go, with only constant gapPen (gap penalty)
+	//bestScore, aln := align.ConstGap(faOne.Seq, faTwo.Seq, align.HumanChimpTwoScoreMatrix, -430)
+	//fmt.Printf("Alignment score is %d, cigar is %v \n", bestScore, aln)
+	//raven's edit: try AffineGap in align/affineGap.go, with diffrent gapOpen and gapExtend penalty, since gapExtend<gapOpen, encourages big gap in order to align better in unbroken chunks
+	bestScore, aln := align.AffineGap(faOne.Seq, faTwo.Seq, align.HumanChimpTwoScoreMatrix, -600, -150)
+	fmt.Printf("Using AffineGap, Alignment score is %d, cigar is %v \n", bestScore, aln)
+
+	//TODO: raven's cigarToBed main body starts here
+	//type definitions: Cigar is defined in align/align.go; aln is a []cigar, according to ConstGap defined in align/linearGap.go
+	//modeled after: mafIndels.go, wigPeaks.go
+	//insertion bed
+	insBed := fileio.EasyCreate(outIns_bed)
+	defer insBed.Close()
+	//TODO: now have to write FirstPos, but find a way to grab from fasta header
+	ChromCurrent := FirstPos_InsBed - 1 //initialize a variable to keep track of current position on chromosome
+	ChromStart := FirstPos_InsBed - 1   //initialize variables for ChromStart and ChromEnd, will need to set them equal to one another during updates, but may be more efficient to have fewer variables
+	ChromEnd := FirstPos_InsBed - 1
+	for i := 0; i < len(aln)-1; i++ { //loop through each entry in the []cigar, start from 0, end at len-2 so can still assess aln[i+1]
+		if aln[i].Op == 0 && aln[i+1].Op == 1 { //if there is an M before I, then write to bed
+			ChromStart = ChromCurrent + int(aln[i].RunLength) + 1                                                       //RunLengths are int64, but bed fields like ChromStart are int, so need to convert RunLengths to int, get the position at which I starts
+			ChromEnd = ChromStart + int(aln[i+1].RunLength)                                                             //get the last position that is I
+			ins := bed.Bed{Chrom: Chrom, ChromStart: ChromStart, ChromEnd: ChromEnd, Name: "ins", FieldsInitialized: 4} //TODO: now just write "chr1", but find a way to grab from fasta header?
```

Comment:
Because you immediately write this bed, you could declare ins before the loop and reuse the struct instead of allocating a new one for each run through the loop.

Reply:
Just fixed. Thank you! I can see how it'll speed things up